pennylane/pytrees.py

```diff
@@ -54,5 +105,93 @@ def register_pytree(pytree_type: type, flatten_fn: FlattenFn, unflatten_fn: Unfl
 
     """
 
+    _register_pytree_with_pennylane(pytree_type, flatten_fn, unflatten_fn)
+
     if has_jax:
         _register_pytree_with_jax(pytree_type, flatten_fn, unflatten_fn)
+
+
+class Structure(namedtuple("Structure", ["type", "metadata", "children"])):
+    """A pytree data structure, holding the type, metadata, and child pytree structures."""
+
+    def __repr__(self):
+        return f"PyTree({self.type.__name__}, {self.metadata}, {self.children})"
+
+
+class Leaf:
+    """A terminal node in a pytree."""
+
+    def __repr__(self):
+        return "Leaf"
+
+    def __eq__(self, other):
+        return isinstance(other, Leaf)
+
+    def __hash__(self):
+        return hash(Leaf)
+
+
+leaf = Leaf()
```

Comment:
```suggestion
class LeafType:
    """Singleton type for a terminal node in a pytree."""
    
    def __new__(cls) -> "LeafType":
        return cls.__instance__
        
    def __repr__(self):
        return "Leaf"

    def __hash__(self):
        return hash(Leaf)

LeafType.__instance__ = object.__new__(LeafType)
Leaf = LeafType()
```
If we implement `Leaf` as a singleton class we can avoid the `isinstance` check in `__eq__` and we also get `None`-like semantics with `is`:

```
>>> x = LeafType()
>>> Leaf is x
True
>>> Leaf == x
True
```

Reply:
Went with considering a leaf a structure with `type is None`. One fewer class to keep track of.